halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/services/v1/ClusterService.java

```diff
@@ -0,0 +1,133 @@
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+
+/**
+ *
+ */
+@Component
+public class ClusterService {
+  @Autowired
+  private LookupService lookupService;
+
+  @Autowired
+  private ProviderService providerService;
+
+  @Autowired
+  private ValidateService validateService;
+
+  @Autowired
+  private OptionsService optionsService;
+
+  public List<Cluster> getAllClusters(String deploymentName, String providerName) {
+    NodeFilter filter = new NodeFilter().setDeployment(deploymentName).setProvider(providerName).withAnyCluster();
+
+    List<Cluster> matchingClusters = lookupService.getMatchingNodesOfType(filter, Cluster.class);
+
+    if (matchingClusters.size() == 0) {
+      throw new ConfigNotFoundException(
+          new ConfigProblemBuilder(Problem.Severity.FATAL, "No clusters could be found").build());
+    } else {
+      return matchingClusters;
+    }
+  }
+
+  private Cluster getCluster(NodeFilter filter, String clusterName) {
+    List<Cluster> matchingClusters = lookupService.getMatchingNodesOfType(filter, Cluster.class);
+
+    switch (matchingClusters.size()) {
+    case 0:
+      throw new ConfigNotFoundException(new ConfigProblemBuilder(
+          Problem.Severity.FATAL, "No cluster with name \"" + clusterName + "\" was found")
+          .setRemediation("Check if this cluster was defined in another provider, or create a new one").build());
+    case 1:
+      return matchingClusters.get(0);
+    default:
+      throw new IllegalConfigException(new ConfigProblemBuilder(
+          Problem.Severity.FATAL, "More than one cluster named \"" + clusterName + "\" was found")
+          .setRemediation("Manually delete/rename duplicate clusters with name \"" + clusterName + "\" in your halconfig file").build());
+    }
+  }
+
+  public Cluster getProviderCluster(String deploymentName, String providerName, String clusterName) {
+    NodeFilter filter = new NodeFilter().setDeployment(deploymentName).setProvider(providerName).setCluster(clusterName);
+    return getCluster(filter, clusterName);
+  }
+
+  public Cluster getAnyProviderCluster(String deploymentName, String clusterName) {
+    NodeFilter filter = new NodeFilter().setDeployment(deploymentName).withAnyProvider().setCluster(clusterName);
+    return getCluster(filter, clusterName);
+  }
+
+  public void setCluster(String deploymentName, String providerName, String clusterName, Cluster newCluster) {
+    Provider provider = providerService.getProvider(deploymentName, providerName);
+    if (!provider.providerType().equals(Provider.ProviderType.DCOS)) {
+      throw new IllegalArgumentException("Provider type does not accept cluster configuration");
+    }
+
+    // TODO(willgorman): could also introduce a ClusterProvider subclass of Provider instead of going straight to DC/OS here
+    DCOSProvider dcosProvider = (DCOSProvider) provider;
```

Comment:
I would prefer if you could switch this to a provider subclass, similar to the `HasImageProvider<T>` so by the time the next person comes around to possibly support a cluster-like construct they aren't tempted to expand these functions into a series of `if (provider ==...)` statements.